### DISCUSSION TITLE
Video length and improving predictions

### DISCUSSION THREAD

**Discussion by IsabelleSajonia** [talmolab/sleap#1656] [General]
Hi, I've been working to generate a sleap model that gives me predictions I can then use in a behavior classifier (SimBA). I had some general questions about best uses. Is there a limitation to the length of videos that you can get inferences for? We give a drug to our mice that changes behavior for ~4 hours but we could break down the video into smaller chunks. I'm also assuming we need to label a lot of frames to get accurate results on a video of this length. Do you have any general recommendations about the amount of labeling needed? I've labeled ~5,000 frames and our model has an OKS mAP of 0.9, which is pretty good, but when I run inference on the whole training video (2 hours) I still get some chunks of frames where the predictions don't look good.

**Comment by Marielamend** [comment]
Hello @IsabelleSajonia!,

Thank you for your inquiry, first of all, labeling 5,000 frames is very impressive! 
To best support you, would you mind sharing some more information about your workflow? If you can share some screenshots of your configurations and datasets that would be great!

Best, 
Mariela

**Comment by IsabelleSajonia** [comment]
![image](https://github.com/talmolab/sleap/assets/70863857/b1284d02-8a52-4868-8a71-8664289d0f94)
Here is a screenshot in the GUI, I mostly labeled frames of a 9 point skeleton in this 2 hour video of a mouse lever pressing but let me know if you wanted something more specific. I will occasionally run inference on the whole video and take note of frames that did not look good to go back and label them, then train again. 
![image](https://github.com/talmolab/sleap/assets/70863857/99f0e61f-4163-413a-8ce5-1246baa79e84)
I mostly just continue training this model using default settings.
![image](https://github.com/talmolab/sleap/assets/70863857/8d381806-d9af-4867-9dc1-2679d1d89916)
This is what I use to run inference, I found that tracking with flow works better for me but otherwise have not tried changing settings much.

**Comment by Marielamend** [comment]
Thank you for sharing @IsabelleSajonia! 

In terms of your configuration there are some things you can modify:
1) Set the rotation augmentation to +-180 since you have an overhead point of view in your video
2) Increase the **max stride** to 32, such that the model is better able to reason about larger spatial scales-- this will increase your receptive field size which should be around the size of the animal that you want to track. (Receptive field represented by blue box on the left of the model configuration window)

Best,
Mariela

**Comment by roomrys** [comment]
Hi @IsabelleSajonia,

Chiming in

>Is there a limitation to the length of videos that you can get inferences for? 

If you are experiencing some slowdown when running inference on extremely long videos, then that could be a combination of both your computer's memory and also the way that SLEAP handles seeking between **labeled** frames. Currently, there is slowdown for interacting with the GUI that scales proportionally to the number of **labeled** frames (this includes frames that are predicted on), so splitting into multiple project could make your user experience a bit better (but I would only recommend if you are experiencing lag).

>I'm also assuming we need to label a lot of frames to get accurate results on a video of this length. Do you have any general recommendations about the amount of labeling needed?

I'm glad you bring this up! It is not the length of the video that (directly) determines the amount of annotations that are needed - instead it is actually the variability of the animal's poses (and positions w/in the frame which changes the lighting on the animals). So yes, in a longer video we will naturally encounter more variability in the poses/lighting and will need more labels - how many labels is open-ended unfortunately and determined by feedback in the model's performance (I'll talk more on this under your next point since you mention the mAP).

>I've labeled ~5,000 frames and our model has an OKS mAP of 0.9, which is pretty good, but when I run inference on the whole training video (2 hours) I still get some chunks of frames where the predictions don't look good.

Whoa that is both a lot of frames and an awesome OKS mAP score! One thing that you should probably know about the OKS score is that it is calculated by measuring how the model performs against ground truth labels (which that specific model was not trained on). However! This also means that if you predict on a frame or chunk of frames that the model is unfamiliar with (has no training data or not enough training data for those poses/lighting/position), then it is totally possible for the model to perform below its expected level (determined by the OKS score).

#### Action items
I would first take a look at the chunks of frames that have poor predictions and try to find similarities/patterns in those frames. How are they different from other frames where the model performs well? Look at animal pose, position in the cage, and lighting. At 5k labels, you would have an extremely strong base (if not full-on pyramid)... adding additional frames tends to have diminishing returns. At a lower frame count I would have suggested to just add more labels to those chunks of frames with poor predictions to include examples for the model. If you notice differences in lighting of the animal, then in addition to @Marielamend's suggestions for the configuration, I think we might be better off trying to augment our already labeled images with playing with some options I've never recommended/tried before (so 🤞🏼 all goes well): brightness and contrast.

Thanks,
Liezl

**Comment by IsabelleSajonia** [comment]
Thank you both, this is really helpful! I'll let you know how this goes!

**Comment by IsabelleSajonia** [comment]
Hi, again. I ran training a few more times with the configuration changes mentioned here. I expected OKS mAP to be low at first, but it doesn't seem to be improving much over multiple training sessions. I'm not sure if I changed something I shouldn't have or if I should make additional modifications. The old configuration was used for the model at the very bottom.
![image](https://github.com/talmolab/sleap/assets/70863857/7f66ff9a-80d2-4255-b9cd-da8497c9185c)
![image](https://github.com/talmolab/sleap/assets/70863857/11480265-1ece-4da9-9064-ceba379ec5bb)

**Reply by Marielamend** [reply]
Hello @IsabelleSajonia,

So sorry for the late response, 
You can try two different approaches for training under your training pipeline:
1. You can try to use the top-down approach and set max stride to 64 to further increase your receptive field.
2. You can also try the bottom-up approach and set max stride to 64, under this approach you can adjust your skeleton and make the `center_3` node the _source_ connecting the other nodes. 

You can experiment with these approaches to see if you get better results!

Give this a try and let us know how it goes!,
Mariela

**Reply by roomrys** [reply]
Looking at your mAP score, it definitely seems like the mAP was really good with a stride of 16 and then changing it to 32 (as was suggested) actually deceased the mAP score by half (0.8 -> 0.4). Conflicting advice, but I would stick to the original configuration based off the results you showed.

**Reply by IsabelleSajonia** [reply]
Hm ok I might try the original configuration again. In my config I also check to stop training on a plateau (to not overfit I think). At the beginning of training a model is it ok to uncheck this? When I first start training the improvements happen very gradually and I'm wondering if we'd train faster by unchecking this and running a lot of epochs overnight for example. Thanks!
